worlds/marioland2/logic.py

```diff
@@ -0,0 +1,608 @@
+
+
+def macro_zone_1_secret_exit(state, player):
+    return state.has("Fire Flower", player) and has_pipe_up(state, player) and macro_zone_1_midway_bell(state, player)
+
+
+def macro_zone_1_midway_bell(state, player):
+    return has_pipe_down(state, player) or state.has("Macro Zone 1 Midway Bell", player)
+
+
+def macro_zone_1_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Macro Zone 1")
+    reachable_coins = 0
+    if has_pipe_down(state, player):
+        reachable_coins += 69
+        if auto_scroll:
+            if state.has_any(["Mushroom", "Fire Flower"], player):
+                reachable_coins += 5
+        else:
+            reachable_coins += 9
+            if state.has("Fire Flower", player):
+                reachable_coins += 19
+    elif state.has("Macro Zone 1 Midway Bell", player):
+        if auto_scroll:
+            reachable_coins += 16
+            if state.has_any(["Mushroom", "Fire Flower"], player):
+                reachable_coins += 5
+        else:
+            reachable_coins += 67
+    return coins <= reachable_coins
+
+
+def macro_zone_secret_course_coins(state, player, coins):
+    return state.has_any(["Mushroom", "Fire Flower"], player)
+
+
+def macro_zone_2_normal_exit(state, player):
+    return (has_pipe_down(state, player) or state.has("Macro Zone 2 Midway Bell", player)) and state.has(
+        "Water Physics", player) and has_pipe_up(state, player) and not is_auto_scroll(state, player, "Macro Zone 2")
+    
+    
+def macro_zone_2_midway_bell(state, player):
+    return ((has_pipe_down(state, player) and state.has("Water Physics", player))
+            or state.has("Macro Zone 2 Midway Bell", player))
+
+
+def macro_zone_2_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Macro Zone 2")
+    if coins <= 27:
+        return True
+    if has_pipe_up(state, player) and state.has("Water Physics", player) and not auto_scroll:
+        if has_pipe_down(state, player):
+            return True
+        if state.has("Macro Zone 2 Midway Bell", player):
+            # Cannot return to the first section from the bell
+            return coins <= 42
+
+
+def macro_zone_3_normal_exit(state, player):
+    return ((has_pipe_down(state, player) and has_pipe_down(state, player))
+            or state.has("Macro Zone 3 Midway Bell", player),)
+
+
+def macro_zone_3_midway_bell(state, player):
+    return ((has_pipe_down(state, player) and has_pipe_down(state, player))
+            or state.has("Macro Zone 3 Midway Bell", player))
+
+
+def macro_zone_3_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Macro Zone 3")
+    reachable_coins = 7
+    if not auto_scroll:
+        reachable_coins = 24
+    if has_pipe_up(state, player) and has_pipe_down(state, player):
+        if auto_scroll:
+            reachable_coins += 56
+        else:
+            return True
```

Comment:
You said you were avoiding this type of construction, so this should use something with `reachable_coins` instead

Reply:
I didn't want to change code to _remove_ information about coin counts. If it's not already there but it's working, it doesn't need any action. The amount of information I made obvious in the code was generally dependent on how complicated the logic for that level was.